From a7d6447e32fed6a583a5313fd375dfdb76518b79 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:05:25 -0400
Subject: [usb] ftdi_sio: added CS5 quirk for broken smartcard readers

Message-id: <1410887463-20674-254-git-send-email-dzickus@redhat.com>
Patchwork-id: 95419
O-Subject: [RHEL7 PATCH 253/591] USB: ftdi_sio: added CS5 quirk for broken smartcard readers
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit c1f15196ac3b541d084dc80a8fbd8a74c6a0bd44
Author: Colin Leitner <colin.leitner@googlemail.com>
Date:   Mon Jan 6 21:33:54 2014 +0100

    USB: ftdi_sio: added CS5 quirk for broken smartcard readers

    Genuine FTDI chips support only CS7/8. A previous fix in commit
    8704211f65a2 ("USB: ftdi_sio: fixed handling of unsupported CSIZE
    setting") enforced this limitation and reported it back to userspace.

    However, certain types of smartcard readers depend on specific
    driver behaviour that requests 0 data bits (not 5) to change into a
    different operating mode if CS5 has been set.

    This patch reenables this behaviour for all FTDI devices.

    Tagged to be added to stable, because it affects a lot of users of
    embedded systems which rely on these readers to work properly.

    Cc: stable <stable@vger.kernel.org>
    Reported-by: Heinrich Siebmanns <H.Siebmanns@t-online.de>
    Tested-by: Heinrich Siebmanns <H.Siebmanns@t-online.de>
    Signed-off-by: Colin Leitner <colin.leitner@gmail.com>
    Signed-off-by: Johan Hovold <jhovold@gmail.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/serial/ftdi_sio.c b/drivers/usb/serial/ftdi_sio.c
index a4bebac..5057201 100644
--- a/drivers/usb/serial/ftdi_sio.c
+++ b/drivers/usb/serial/ftdi_sio.c
@@ -2121,10 +2121,20 @@ static void ftdi_set_termios(struct tty_struct *tty,
  }
 
  /*
-  * All FTDI UART chips are limited to CS7/8. We won't pretend to
+  * All FTDI UART chips are limited to CS7/8. We shouldn't pretend to
   * support CS5/6 and revert the CSIZE setting instead.
+  *
+  * CS5 however is used to control some smartcard readers which abuse
+  * this limitation to switch modes. Original FTDI chips fall back to
+  * eight data bits.
+  *
+  * TODO: Implement a quirk to only allow this with mentioned
+  *       readers. One I know of (Argolis Smartreader V1)
+  *       returns "USB smartcard server" as iInterface string.
+  *       The vendor didn't bother with a custom VID/PID of
+  *       course.
   */
- if ((C_CSIZE(tty) != CS8) && (C_CSIZE(tty) != CS7)) {
+ if (C_CSIZE(tty) == CS6) {
   dev_warn(ddev, "requested CSIZE setting not supported\n");
 
   termios->c_cflag &= ~CSIZE;
@@ -2171,6 +2181,9 @@ no_skip:
   urb_value |= FTDI_SIO_SET_DATA_PARITY_NONE;
  }
  switch (cflag & CSIZE) {
+ case CS5:
+  dev_dbg(ddev, "Setting CS5 quirk\n");
+  break;
  case CS7:
   urb_value |= 7;
   dev_dbg(ddev, "Setting CS7\n");
-- 
1.7.1